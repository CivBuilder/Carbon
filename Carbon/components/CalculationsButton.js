import { View, Text, StyleSheet, Pressable } from 'react-native'
import React from 'react'


const CalculationsButton = ({ onPress }) => {
    return (
        <View >
            <Pressable style={styles.button} onPress={onPress} testID="calculation-details">
                <Text style={styles.buttonText}>Calculation Details</Text>
            </Pressable >
        </View>

    )
}

const styles = StyleSheet.create({
    button: {
        width: 200,
        height: 40,
        backgroundColor: '#74C69D',
        borderRadius: 20,
        alignItems: 'center',
        justifyContent: 'center',
        marginTop: 10,
        marginBottom: 30,
    },
    buttonText: {
        color: 'white',
        fontFamily: 'sans-serif',
    },
})
export default CalculationsButton;