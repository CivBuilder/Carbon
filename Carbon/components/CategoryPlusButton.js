import { StyleSheet, Text, TextInput, View, Pressable, Modal } from 'react-native'
import {React, useState} from 'react'
import SaveEmissions from './SaveEmissions';

const CategoryPlusButton = () => {
  const [modalVisible, setModalVisible] = useState(false);
  const [foodVisible, setFoodVisible] = useState(false);
  const [transportationVisible, setTransportationVisible] = useState(false);
  const [recyclingVisible, setRecyclingVisible] = useState(false);

  
  //TODO: Sanitize and save input to database
  
  const saveFoodLog = (log) => {
    console.log('saving food ' + log);
  }
  const saveTransportationLog = (log) => {
    console.log('saving transportation ' + log);
  }

  const saveRecyclingLog = (log) => {
    console.log('saving recycling ' + log);
  }
  return (
    <View>
        <Modal style={styles.modal}
        animationType="slide"
        transparent={true}
        backdropOpacity={0}
        visible={modalVisible}
        onRequestClose={() => {
          setModalVisible(!modalVisible);
        }}>
        <View style={styles.centeredView}>
          <View style={styles.modalView}>

            <Text style={styles.modalText}>Select the category to log</Text>
            <Pressable onPress={() => setFoodVisible(true)}>
              <Text style={styles.text}>Food</Text>
            </Pressable>

            <Pressable onPress={() => setTransportationVisible(true)}>
              <Text style={styles.text}>Transportation</Text>
            </Pressable>

            <Pressable onPress={() => setRecyclingVisible(true)}>
              <Text style={styles.text}>Recycling</Text>
            </Pressable>

            <Pressable
              style={[styles.button]}
              onPress={() => setModalVisible(!modalVisible)}>
              <Text style={styles.text}>close</Text>
            </Pressable>

          </View>
        </View>
        </Modal>

      <SaveEmissions visible={foodVisible}
                     onRequestClose={() => {
                      setFoodVisible(!foodVisible)
                    }}  
                    title={"How much meat did you consume today?"}
                    saveData={(log) => saveFoodLog(log)}
                    />
        <SaveEmissions visible={transportationVisible}
                     onRequestClose={() => {
                      setTransportationVisible(!transportationVisible)
                    }}  
                    title={"How many miles did you drive today?"}
                    saveData={(log) => saveTransportationLog(log)}
                    />
        <SaveEmissions visible={recyclingVisible}
                     onRequestClose={() => {
                      setRecyclingVisible(!recyclingVisible)
                    }}  
                    title={"How much did you recycle today?"}
                    saveData={(log) => saveRecyclingLog(log)}
                    />

        <Pressable style={styles.button} onPress={() => setModalVisible(true)}>
        <Text style={styles.text}>Add</Text>
        </Pressable>
    </View>
    
  )
}

export default CategoryPlusButton

const styles = StyleSheet.create({
    centeredView: {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center',
        marginTop: 22,
        justifyContent: 'flex-end',
        marginBottom: 48 ,
      },

    modalView: {
        width: "100%",
        height: "33%",
        borderTopStartRadius: 12,
        borderTopEndRadius: 12,
        alignItems: "center",
        justifyContent: "center",
        backgroundColor: '#465966',
        shadowOpacity: 0.25,
        shadowRadius: 4,
        elevation: 5,
      },
    modal: {
      backgroundColor: '#465966',
    },  
    text: {
      textAlign: 'center',
    },
    button: {
        borderRadius: 4,
        borderWidth: 2,
        borderColor: 'black',
        padding: 10,
        margin: 10,
        elevation: 2,
        width: 120,
      },

      modalText: {
        marginBottom: 15,
        textAlign: 'center',
      },
})