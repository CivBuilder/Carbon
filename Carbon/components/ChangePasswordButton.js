import { View, Text, StyleSheet, Pressable } from 'react-native'
import React from 'react'


const ChangePasswordButton = ({disabled, onPress }) => {
    return (
        <View >
            {disabled ? (
                <Pressable style={{...styles.button, backgroundColor: '#C2C2C2'}} onPress={onPress} testID="changePasswordButton" disabled={disabled}>
                    <Text style={styles.buttonText}>Change Password</Text>
                </Pressable >
            ) : (
                <Pressable style={styles.button} onPress={onPress} testID="changePasswordButton">
                    <Text style={styles.buttonText}>Change Password</Text>
                </Pressable >
            )}
        </View>

    )
}

const styles = StyleSheet.create({
    button: {
        width: 200,
        height: 40,
        backgroundColor: '#74C69D',
        borderRadius: 20,
        alignItems: 'center',
        justifyContent: 'center',
        marginTop: 10,
        marginBottom: 30,
    },
    buttonText: {
        color: 'white',
        fontSize: 16,
        fontFamily: 'sans-serif',
    },
})

export default ChangePasswordButton