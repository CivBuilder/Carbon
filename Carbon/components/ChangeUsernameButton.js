import { View, Text, StyleSheet, Pressable } from 'react-native'
import React from 'react'

const ChangeUsernameButton = ({ onPress }) => {
    return (
        <View >
            <Pressable style={styles.button} onPress={onPress} testID="changeUsernameButton">
                <Text style={styles.buttonText}>Change Username</Text>
            </Pressable >
        </View>
    )
}

const styles = StyleSheet.create({
    button: {
        width: 200,
        height: 40,
        backgroundColor: '#74C69D',
        borderRadius: 20,
        alignItems: 'center',
        justifyContent: 'center',
        marginTop: 10,
        marginBottom: 30,
    },
    buttonText: {
        color: 'white',
        fontFamily: 'sans-serif',
    },
})

export default ChangeUsernameButton