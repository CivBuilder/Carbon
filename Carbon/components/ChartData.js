import React, { useState } from 'react';
import { View, Text, StyleSheet, Dimensions } from 'react-native';
import { VictoryPie } from 'victory-native';
import { Colors } from '../colors/Colors';

const windowWidth = Dimensions.get("window").width;
const windowHeight = Dimensions.get("window").height;
const margin = 10;
const chartWidth = windowWidth - (margin * 2);
const chartHeight = 210;

const monthNames = [
    "January", "February", "March", "April", "May", "June", "July",
    "August", "September", "October", "November", "December"
];
const today = new Date();
const currentMonth = monthNames[today.getMonth()];

const data = [
    { x: "Transport", y: Math.round(Math.random() * 10000) },
    { x: "Diet", y: Math.round(Math.random() * 10000) },
    { x: "Home", y: Math.round(Math.random() * 10000) },
    { x: "Stuff", y: Math.round(Math.random() * 10000) },
];

// Ticket C4-19
export const CatgegoryChart = () => {
    const total = data.reduce((acc, datum) => acc + datum.y, 0); //Gets the overall value of all categories

    //Components for the pie chart
    const pieRadius = (windowWidth - (margin * 2)) * 0.4;
    const innerRadius = pieRadius * 0.65;
    const labelRadius = ((pieRadius + innerRadius) / 2);

    // Renders the label for each section of the pie chart
    const getLabel = (datum) => {
        const percent = Math.round((datum.y / total) * 100);
        if (percent > 4) {
            return `${datum.x}\n${Math.round((datum.y / total) * 100)}%`;
        }
        return null;
    };

    const [selectedSlice, setSelectedSlice] = useState(null);

    const handlePress = (event, props) => {
        const selectedDatum = data[props.index];
        setSelectedSlice(props.index);
        // console.log(`Selected slice: ${selectedDatum.x} (${selectedDatum.y})`);
    };

    // Shows the value of the selected section in the middle of the pie chart
    const getSelectedLabel = () => {
        if (selectedSlice !== null) {
            const selectedDatum = data[selectedSlice];
            return `${selectedDatum.x}\n${selectedDatum.y} lbs CO2`;
        }
        return null;
    };

    return (
        <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center', marginVertical: -40 }}>
            <VictoryPie
                data={data}
                colorScale={["#5D5FEF", "#FF8C00", "#3CB371", "#FF69B4"]}
                padAngle={2}
                radius={pieRadius}
                innerRadius={innerRadius}
                cornerRadius={6}
                labelRadius={labelRadius}
                labels={({ datum }) => getLabel(datum)}
                style={{
                    labels: {
                        fill: "white",
                        fontSize: 12,
                        fontWeight: "bold",
                        textAnchor: "middle",
                        verticalAnchor: "middle",
                    },
                }}
                events={[{
                    target: 'data',
                    eventHandlers: {
                        onPressIn: handlePress,
                    },
                },]}
                animate={{ duration: 500, ease: "exp" }}
                selected={[selectedSlice]}
            />
            <View
                style={{
                    justifyContent: 'center',
                    alignItems: 'center',
                    position: 'absolute',
                }}>
                <View style={{ paddingBottom: 5 }}>
                    <Text style={{ textAlign: 'center', fontSize: 20, fontWeight: 'bold', textDecorationLine: 'underline' }}>{currentMonth}</Text>
                </View>

                {selectedSlice == null && ( // A slice has not yet selected
                    <View style={{ }}>
                        <Text style={{ textAlign: 'center' }}>Click a section to learn more</Text>
                    </View>
                )}
                {selectedSlice !== null && ( // A slice has been selected
                    <View style={{ }}>
                        <Text style={{ textAlign: 'center', fontSize: 16 }}>{getSelectedLabel()}</Text>
                    </View>
                )}
            </View>
        </View>
    );
};

//Data for later
/*
    The daily log function, this takens in a data array and maps it into a bar chart based on its values
 */
export function DailyLog ({dataArray}) {

    //Our data for bar styling the bar chart
    const barChartData = {
        labels: ['Transportation', 'Diet', 'Lifestyle', 'Home', 'Overall'],
        datasets:[
            {
            data: dataArray,
            colors: [
                (opacity = 1) =>  Colors.primary.MINT,
                (opacity = 1) =>  Colors.primary.MINT,
                (opacity = 1) =>  Colors.primary.MINT, //all colors will be MINt
                (opacity = 1) =>  Colors.primary.MINT,
                (opacity = 1) =>  Colors.primary.MINT,
            ]
            }
        ]
        };
        //Return the bar chart with appropriate styling.
    return (
        <BarChart
        data ={barChartData}
        chartConfig={styleBar.chartConfig}
        withCustomBarColorFromData={true}
        flatColor={true}
        style={styleBar.chart}
        width = {chartWidth}
        height = {windowHeight/3}
        withVerticalLabels ={true}
        withHorizontalLabels = {true}
        fromZero = {true}
    />
    )
}

const styles = StyleSheet.create({
    chart: {
        borderRadius: 16,
    },
    chartConfig: {
        backgroundColor: Colors.primary.MINT,
        backgroundGradientFrom: Colors.primary.MINT,
        backgroundGradientTo: Colors.primary.MINT,
        decimalPlaces: 2, // optional, defaults to 2dp
        color: (opacity = 1) => `rgba(255, 255, 255, ${opacity})`,
        labelColor: (opacity = 1) => `rgba(255, 255, 255, ${opacity})`,
        propsForDots: {
            r: "5",                         // circle size
            strokeWidth: "2",               // circle border size
            stroke: Colors.primary.MINT     // circle border color
        }
    },
});

//Style bar which was created by Miguel
const styleBar = StyleSheet.create({
    container: {
        marginLeft: 100,
        flex: 1,
        justifyContent: 'center', padding: 100,
        paddingTop: 30, backgroundColor: '#ecf0f1',
        },
    chart: {
        borderRadius: 16,
        marginHorizontal: margin/2,
    },
    chartConfig: {

        backgroundColor: '#FFFFFF',
        backgroundGradientFrom: '#FFFFFF',
        backgroundGradientTo: '#FFFFFF',
        decimalPlaces: 2, // optional, defaults to 2dp
        color: (opacity = 1) => `rgba(255, 255, 255, ${opacity})`, //transparent
        labelColor: (opacity = 1) => `rgba(0, 0, 0, ${opacity})`,

        propsForDots: {
            r: "5",                         // circle size
            strokeWidth: "2",               // circle border size
            stroke: Colors.primary.MINT     // circle border color
        }
    },
    keyFactors: {
        flex: 1,
        flexDirection: 'column',
        justifyContent: 'space-between',
        alignItems: 'center',
        marginHorizontal: margin,
        marginVertical: margin-5
    },
    category: {
        flexDirection: 'row',
        justifyContent: 'space-between',
        alignItems: 'center',
        flexWrap: 'nowrap',
        marginVertical: 5
    },
});