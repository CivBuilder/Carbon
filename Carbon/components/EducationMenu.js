import {View, Text, TouchableOpacity, Image, StyleSheet} from 'react-native'
import { Colors } from '../styling/Colors'
import React from 'react'


const EducationMenu = ({title, imageSrc, onPress}) => {
    const test = {
        backgroundColor: '#43b262'
    }
    return(
        <View style={{paddingHorizontal: 10, alignItems: 'center'}}>
            <TouchableOpacity style = {styles.roundButton}  onPress={onPress}>
                <Image style = {styles.categoryIcon} source={imageSrc} />
            </TouchableOpacity>
            <Text style = {styles.imageText}>{title}</Text>
        </View>
   )
}

const styles = StyleSheet.create({
    roundButton: {
        height: 70,
        width: 70,
        justifyContent: 'center',
        alignItems: 'center',
        borderRadius: 100,
        backgroundColor: Colors.primary.MINT_CREAM,
        borderColor: Colors.primary.MINT,
        borderWidth: 3,
    },

    categoryIcon: {
        resizeMode: 'contain',
        height: '80%',
        width: '80%',
        tintColor: Colors.primary.MINT,
    },

    imageText: {
        fontSize: 17,
        fontWeight: 'bold',
        color: 'gray',
        textAlign: 'center'
    }
});


export default EducationMenu
