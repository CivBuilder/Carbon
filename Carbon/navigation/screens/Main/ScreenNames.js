export const ScreenNames = {
    HOME: 'Home',
    PROGRESS: 'Progress',
    FORUM: 'Forum',
    RANKING: 'Ranking',
    SETTINGS: 'Settings',
    ADD_PROGRESS: 'AddProgress',
    QUIZSCREEN: 'Quiz',
};