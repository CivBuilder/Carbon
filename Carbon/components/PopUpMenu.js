import React, { useState } from 'react';
import { View, TouchableOpacity, Modal, Text, StyleSheet, Platform } from 'react-native';
import { Ionicons } from '@expo/vector-icons';

import { IconNames } from '../navigation/screens/Main/IconNames';
import { ScreenNames } from '../navigation/screens/Main/ScreenNames';
import { Colors } from '../colors/Colors';


export const PopUpMenu = ({navigation}) => {
    /*  Handles the visibility of the pop-up menu when the plus button is pressed,
        when user pressed anywhere outside of the modal, and when user clicks one of the options
    */
    const [visible, setVisibility] = useState(false);

    // Lists out the options user can do within this pop-up menu
    const options = [
        {
            title: 'Record Emissions',
            icon: 'checkmark-circle-outline',
            action: () => {
                console.log("Record Emissions button is clicked"); // TODO: Connect to the Record Emissions page
                // TODO: Add the records emissions page in ProgressStack in MainContainer.js
            }
        },
        {
            title: 'Add Goal',
            icon: 'timer-outline',
            action: () => {
                navigation.navigate(ScreenNames.ADD_GOAL);
            },
        },
    ];

    return (
        <View>
            {/* Displays the pop-up menu when the + button is pressed */}
            <TouchableOpacity
                onPress={() => setVisibility(true)}
            >
                <Ionicons
                    name={IconNames.ADD}
                    size={28}
                    color={Colors.primary.MINT}
                    style={{ marginRight: 16 }}
                />
            </TouchableOpacity>

            <Modal transparent visible={visible}>
                {/* Clicking outside of the modal will close the menu */}
                {/* NOTE: Keep the backdrop style in. Removing it causes the modal to not disappear when clicking outside */}
                <TouchableOpacity style={styles.backdrop} onPress={() => setVisibility(false)}/>

                {/* Handles the menu inside the modal */}
                <View style={styles.popup_modal}>
                    {options.map((op, i) => (
                        <TouchableOpacity
                            style={[styles.options, { borderBottomWidth: i === options.length - 1 ? 0 : 1 }]}
                            key={i}
                            onPress={() => {
                                op.action();
                                setVisibility(false);
                            }}
                        >
                            {/* Display the menu text */}
                            <Text style={styles.popup_title}>{op.title}</Text>

                            {/* Display the menu icons */}
                            <Ionicons
                                name={op.icon}
                                size={24}
                                color={Colors.primary.RAISIN_BLACK}
                            />
                        </TouchableOpacity>
                    ))}
                </View>
            </Modal>
        </View>
    );
}

const styles = StyleSheet.create({
    popup_modal: {
        borderRadius: 8,
        borderWidth: 1,
        borderColor: Colors.primary.RAISIN_BLACK,
        backgroundColor: "white",
        paddingHorizontal: 16,
        position: 'absolute',
        right: 24,
        top: Platform.OS === 'ios' ? 69 : 43, // TODO: Use `useHeaderHeight()` hook to get the height of the header
    },
    popup_title: {
        paddingRight: 16,
    },
    options: {
        flexDirection:'row',
        justifyContent: 'space-between',
        alignItems: 'center',
        paddingVertical: 8,
        borderBottomColor: '#ccc',
    },
    backdrop: { //NOTE: Keep the backdrop style in. Removing it causes the modal to not disappear when clicking outside
        position: 'absolute',
        top: 0,
        left: 0,
        right: 0,
        bottom: 0,
    },
});