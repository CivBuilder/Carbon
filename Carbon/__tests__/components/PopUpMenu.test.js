import React from "react";
import { render, fireEvent } from "@testing-library/react-native";
import { create } from 'react-test-renderer';
import { PopUpMenu } from "../../components/PopUpMenu";
import { ScreenNames } from "../../navigation/screens/Main/ScreenNames";

describe("PopUpMenu", () => {
    test('renders correctly', () => {
        const tree = create(<PopUpMenu />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('displays the pop-up menu when the + button is pressed', () => {
        const navigation = { navigate: jest.fn() };
        const { getByTestId } = render(<PopUpMenu navigation={navigation} />);
        const plusButton = getByTestId('plus_button');

        fireEvent.press(plusButton);

        expect(getByTestId('popup_menu')).toBeVisible();
    });

    it('closes the pop-up menu when the backdrop is pressed', () => {
        const navigation = { navigate: jest.fn() };
        const { getByTestId } = render(<PopUpMenu navigation={navigation} />);
        const plusButton = getByTestId('plus_button');

        fireEvent.press(plusButton);
        fireEvent.press(getByTestId('popup_menu_backdrop'));

        expect(getByTestId('popup_menu')).not.toBeVisible();
    });

    test('renders plus button', () => {
        const { getByTestId } = render(<PopUpMenu navigation={jest.fn()} />);
        expect(getByTestId('plus_button')).toBeTruthy();
    });

    test('renders menu options', () => {
        const { getByTestId, getByText } = render(<PopUpMenu navigation={jest.fn()} />);
        fireEvent.press(getByTestId('plus_button'));

        expect(getByTestId('popup_menu')).toBeTruthy();
        expect(getByText('Record Emissions')).toBeTruthy();
        expect(getByText('Add Goal')).toBeTruthy();
    });

    test('navigates to correct screen when option is pressed', () => {
        const navigateMock = jest.fn();
        const { getByTestId, getByText } = render(<PopUpMenu navigation={{ navigate: navigateMock }} />);
        fireEvent.press(getByTestId('plus_button'));
        fireEvent.press(getByText('Record Emissions'));

        expect(navigateMock).toHaveBeenCalledWith(ScreenNames.RECORD_EMISSION);
    });

    it('navigates to the record emissions screen when the "Record Emissions" option is selected', () => {
        const navigateMock = jest.fn();
        const { getByTestId, getByText } = render(<PopUpMenu navigation={{ navigate: navigateMock }} />);
        fireEvent.press(getByTestId('plus_button'));
        fireEvent.press(getByText('Record Emissions'));
        expect(navigateMock).toHaveBeenCalledWith(ScreenNames.RECORD_EMISSION);
        expect(getByTestId('popup_menu')).not.toBeVisible();
    });

    it('navigates to the add goal screen when the "Add Goal" option is selected', () => {
        const navigateMock = jest.fn();
        const { getByTestId, getByText } = render(<PopUpMenu navigation={{ navigate: navigateMock }} />);
        fireEvent.press(getByTestId('plus_button'));
        fireEvent.press(getByText('Add Goal'));
        expect(navigateMock).toHaveBeenCalledWith(ScreenNames.ADD_GOAL);
        expect(getByTestId('popup_menu')).not.toBeVisible();
    });
});