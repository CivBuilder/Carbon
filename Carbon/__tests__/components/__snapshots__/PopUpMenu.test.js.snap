// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopUpMenu renders correctly 1`] = `
<View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Text />
  </View>
  <Modal
    hardwareAccelerated={false}
    testID="popup_menu"
    transparent={true}
    visible={false}
  />
</View>
`;
